### PR TITLE
New sketch, DS3231_Temperature.ino

### PR DESCRIPTION
**Pull request by mrwgx3**
Please reference issue #49 and pull-request #51

I'm proposing that the sketch 'DS3231_Temperature.ino' be added to 'examples' to illustrate the use of the updated and new methods found in class RtcTemperature. This sketch was derived from 'DS3231_Simple.ino', and also includes minor typo and code corrections.

### CODE REVIEW

**Comment by mrwgx3** [comment]
Here are the corresponding changes for wiki file 'RtcTemperature object', both in raw andMarkdown format, which describe the methods tested:

### Raw format: 
```
RtcTemperature represents the temperature one object.  It is used to retrieve the temperature


# Constructors

### RtcTemperature(int8_t msdegC, uint8_t lsdegC)  
Construct a temperature object from a scaled (x 256) integer temperature.  
_msdegC_ - MS byte  
_lsdegC_ - LS byte, bounded to 0.25 degC resolution  

# Methods
### int16_t  AsScaledDegrees()
\<return\>, scaled (x 256) temperature Celsius ( -23.25 scales to -5952 ) 

### int8_t  AsRoundedDegrees()
\<return\>, the rounded temperature Celsius ( +/-23.50 rounds to  +/-24 )  

### float  AsFloat()  
\<return\>, the temperature Celsius as a float ( 23.75 )  
   
### int8_t  AsWholeDegrees()
\<return\>, whole degrees Celsius ( +/-23.xx returns +/-23 ) 

### uint8_t  GetFractional() 
\<return\>, scaled fractional degrees Celsius ( +/-23.xx returns 0, 25, 50, or 75 )  
``` 


### Markdown format:

RtcTemperature represents the temperature one object.  It is used to retrieve the temperature


# Constructors

### RtcTemperature(int8_t msdegC, uint8_t lsdegC)  
Construct a temperature object from a scaled (x 256) integer temperature.  
_msdegC_ - MS byte  
_lsdegC_ - LS byte, bounded to 0.25 degC resolution  

# Methods
### int16_t  AsScaledDegrees()
\<return\>, scaled (x 256) temperature Celsius ( -23.25 scales to -5952 ) 

### int8_t  AsRoundedDegrees()
\<return\>, the rounded temperature Celsius ( +/-23.50 rounds to  +/-24 )  

### float  AsFloat()  
\<return\>, the temperature Celsius as a float ( 23.75 )  
   
### int8_t  AsWholeDegrees()
\<return\>, whole degrees Celsius ( +/-23.xx returns +/-23 ) 

### uint8_t  GetFractional() 
\<return\>, scaled fractional degrees Celsius ( +/-23.xx returns 0, 25, 50, or 75 )

**Comment by mrwgx3** [comment]
Here are the corresponding changes for wiki file 'Examples', both in Markdown and raw format, which describe the new sketch:

Raw format: 
```
You can find these examples in the Arduino IDE examples menu under "Rtc by Makuna".

### DS1307 Simple  
This demonstrates how to set and get the date and time from the device.

### DS1307 Memory  
This demonstrates how to write and read data from the extended memory of  
the device. This memory is retained even after the device has lost power.

### DS3231 Simple  
This demonstrates how to set and get the date and time from the device.

### DS3231 StoreIt  
This demonstrates how to turn off the Rtc clock so that the  
onboard battery will last longer.

### DS3231 Alarms  
This demonstrates how to set the alarms and check when the alarms are  
triggered. This includes using the squarewave output to trigger an interrupt.

### DS3231 Temperature  
This demonstrates how to initiate a temperature conversion,  
then retrieve and print the result in various formats.

```

Markdown format:

You can find these examples in the Arduino IDE examples menu under "Rtc by Makuna".

### DS1307 Simple  
This demonstrates how to set and get the date and time from the device.

### DS1307 Memory  
This demonstrates how to write and read data from the extended memory of  
the device. This memory is retained even after the device has lost power.

### DS3231 Simple  
This demonstrates how to set and get the date and time from the device.

### DS3231 StoreIt  
This demonstrates how to turn off the Rtc clock so that the  
onboard battery will last longer.

### DS3231 Alarms  
This demonstrates how to set the alarms and check when the alarms are  
triggered. This includes using the squarewave output to trigger an interrupt.

### DS3231 Temperature  
This demonstrates how to initiate a temperature conversion,  
then retrieve and print the result in various formats.

File: examples/DS3231_Temperature/DS3231_Temperature.ino [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+#include <RtcDS3231.h>
+RtcDS3231<TwoWire> Rtc(Wire);
+/* for normal hardware wire use above */
+
+
+void setup () 
+{
+    Serial.begin(57600);
+    while (!Serial) delay(250);    // Wait until Arduino Serial Monitor opens
+    Serial.println( " " );         // For ESP8266, newline after bootup message
```
Comment (Makuna) [inline examples/DS3231_Temperature/DS3231_Temperature.ino]: don't include platform specific comments/code in examples as it makes them harder to read.
remove the comment at least; also, `Serial.println()` is good enough, no need to print the space.

File: examples/DS3231_Temperature/DS3231_Temperature.ino [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+{
+    if (!Rtc.IsDateTimeValid()) 
+    {
+        // Common Causes:
+        //    1) the battery on the device is low or even missing and the power line was disconnected
+        Serial.println("RTC lost confidence in the DateTime!");
+    }
+
+    // Force a temperature A/D conversion
+    Rtc.ForceTemperatureCompensationUpdate( true );
```
Comment (Makuna) [inline examples/DS3231_Temperature/DS3231_Temperature.ino]: Using the method "ForceTemperatureCompenstationUpdate()" should be avoided in most cases.  It is not needed to get an accurate temperature. 

All it does is force the chip to change how it compensates the passage of time based on temperature to be reevaluated, which it already does periodically.  Calling this is more important if you want a "VERY" accurate time in a "VERY" fast changing temperature environment.

Were you finding this a requirement to have the temperatures change at all?  This is not the case for earlier RTC chips.

File: examples/DS3231_Temperature/DS3231_Temperature.ino [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+{
+    // Get the temperature data    
+    RtcTemperature temp = Rtc.GetTemperature();   
+
+    // Print:
+    // a) Floating point temperature
+    //    Selectively add spaces to create const. width field
+    //    ( no float formating for 'printf' type functions )
+
+    float  tf = temp.AsFloat();
```
Comment (Makuna) [inline examples/DS3231_Temperature/DS3231_Temperature.ino]: separate the print float and print fractional into their own functions, makes this easier to read for beginners.

File: examples/DS3231_Temperature/DS3231_Temperature.ino (hunk 2 above)
Comment (mrwgx3) [reply examples/DS3231_Temperature/DS3231_Temperature.ino]: Remember that 'DS3231_Temperature.ino' is intended to test the methods included in RtcTemperature. The "ForceTemperatureCompenstationUpdate()" is included to show its used, and a rapid temperature update is 
desired here to show heating/cooling of the part.

Have added a comment stating the above to the sketch

File: examples/DS3231_Temperature/DS3231_Temperature.ino (hunk 1 above)
Comment (mrwgx3) [reply examples/DS3231_Temperature/DS3231_Temperature.ino]: Will remove '#if defined(ESP8266)' / '#endif'
Regarding Serial.println( " " )...
I'm using the both the Huzzah breakout and feather boards from Adafruit, which use the ESP12 part. Upon bootup, the ESP12 sends a long message at 74880 baud, which generates overrun/framing errors for the connected serial port when its baud-rate doesn't match. As a result, the newline character sent by a 'Serial.println()' is occasionally  missed, 
Inserting the space clears these errors, insuring the subsequent newline character is seen.

I've left the Serial.println( " " ) intact, but have removed the comment.

File: examples/DS3231_Temperature/DS3231_Temperature.ino (hunk 3 above)
Comment (mrwgx3) [reply examples/DS3231_Temperature/DS3231_Temperature.ino]: Have separated them out using 'Serial.printf()' statements. A new method, 'IsMinusSignNeeded()', has been added to get around a conceptual error regarding 'AsWholeDegrees()' and 'GetFractional()'. See pull request (#51) for more details.

File: examples/DS3231_Temperature/DS3231_Temperature.ino [hunk 4]
```diff
@@ -129,39 +154,55 @@ void printDateTime(const RtcDateTime& dt)
...
     // Print:
-    // b) Whole degrees/fractional portions
-    //    formatted as a float
+    // b) Whole degrees/fractional portions formatted as a float
+    //     i) 'printf' has no problems with integers
+    //    ii) Special test needed -0.75 degC through -0.25 degC
+
+    Serial.printf(
+       // Distinguish between:        -0.xx   or   +0.xx       
+       ( td.IsMinusSignNeeded() ? " -%1d.%02u" : "%3d.%02u" ),
```
[the remark this answers is not in the document]
Comment (Makuna) [reply examples/DS3231_Temperature/DS3231_Temperature.ino]: I understand now, ignore my comment above.

**Pull request closed without merging**